### PR TITLE
ENH: Integrate Sequences extension into Slicer core

### LINKED ISSUE #4461: Move Sequences extension to Slicer core
_This issue was created automatically from an original [Mantis Issue](https://mantisarchive.slicer.org/view.php?id=4461). Further discussion may take place here._

### PR DESCRIPTION
**Pull request by lassoan**
Move latest master version of Sequences extension (https://github.com/SlicerRt/Sequences) to Slicer core.
This extension allows recording and replaying of sequences of almost any node types.

Merge Sequences and Sequence Browser modules to make all important sequence features
in one place. Some features, such as plotting may be moved out to separate modules in the future.

Modules and plugins added:
* Sequences (C++ loadable)
* CropVolumeSequence (python scripted): Crop and resample a volume
  sequence to reduce its size for faster rendering and processing
* DICOMVolumeSequencePlugin: Support export of sequence node
  that contains volumes to DICOM
* SampleData: CTPCardioSeq and CTCardioSeq

Design changes made along with the migration:

* PR #4830: Refactor MRML node API to support shallow and deep copy
* PR #4831: Update MRML node API to allow listing of node content modification events
* PR #4832: Ensure vtkCodedEntryTest1 and vtkOrientedGridTransformTest1 are executed.
* PR #4833: Refactor vtkArchive to support use from MRML Core classes.
* PR #4834: Update MRML scene API adding support for read/write of MRML bundle.

Update vtkMRMLNodeTest1:
* Disable test of display node reference
  by introducing DISPLAY_SCENE_NODE_REFERENCES define set to 0

Disabled SceneView tests (they will be fixed in follow-up commits):
* vtkMRMLSceneViewNodeImportSceneTest
* vtkMRMLSceneViewNodeEventsTest
* vtkMRMLSceneViewNodeRestoreSceneTest
* vtkMRMLSceneViewNodeStoreSceneTest
* AnnotationsTestingFiducialWithSceneViewRestore
* MarkupsSceneViewRestoreTestSimple
* MarkupsSceneViewRestoreTestManyLists
* AddStorableDataAfterSceneViewTest

Notes:
* Sequence file reader was moved to SlicerIGT. See SlicerIGT/SlicerIGT#196

Fixes #4461

### CODE REVIEW

**Comment by jcfr** [comment]
Thanks for the contribution 🙏 

I will be able to review tomorrow.

**Comment by hjmjohnson** [comment]
This seems like it is ready for rebasing/merging 5 days ago.

**Comment by jcfr** [comment]
I started the review, we should have this integrated before the end of the week.

**Comment by jcfr** [comment]
Improvements to MRML Node copy API have been pushed into their own PR and will be reviewed there. See #4830

File: Libs/MRML/Core/vtkMRMLDiffusionImageVolumeNode.h [hunk 1]
```diff
@@ -50,7 +50,7 @@ class VTK_MRML_EXPORT vtkMRMLDiffusionImageVolumeNode : public vtkMRMLTensorVolu
 
   ///
   /// Get node XML tag name (like Volume, Model)
-  const char* GetNodeTagName() override {return "DiffusionImageVolume";}
+  const char* GetNodeTagName() override {return "vtkMRMLDiffusionImageVolumeNode";}
```
Comment (jcfr) [inline Libs/MRML/Core/vtkMRMLDiffusionImageVolumeNode.h:53]: Why this was changed ? 

None of the existing tag name corresponds to the class name:

```
$ ack --cpp "GetNodeTagName\("
[...]
Libs/MRML/Core/vtkMRMLDisplayableHierarchyNode.h
53:  const char* GetNodeTagName() override {return "DisplayableHierarchy";}

Libs/MRML/Core/vtkMRMLInteractionNode.h
25:  const char* GetNodeTagName() override {return "Interaction";}

Libs/MRML/Core/vtkMRMLdGEMRICProceduralColorNode.h
52:  const char* GetNodeTagName() override {return "dGEMRICProceduralColor";}

Libs/MRML/Core/vtkMRMLHierarchyStorageNode.h
36:  const char* GetNodeTagName() override {return "HierarchyStorage";}

Libs/MRML/Core/vtkMRMLFiducialListNode.h
74:  const char* GetNodeTagName() override {return "FiducialList";}
[...]
```

**Comment by jcfr** [comment]
@lassoan 

Now all dependent PR have been integrated, this is ready for final review. 

Pending questions:
* what should be the value returned by `vtkMRMLDiffusionImageVolumeNode::GetNodeTagName()`. See https://github.com/Slicer/Slicer/pull/4810#discussion_r409328761
* should sceneview changes be integrated in this PR ?

File: Libs/MRML/Core/vtkMRMLDiffusionImageVolumeNode.h (hunk 1 above)
Comment (lassoan) [reply Libs/MRML/Core/vtkMRMLDiffusionImageVolumeNode.h:53]: Good catch. This change was definitely not intentional and the node tag name should be reverted to the original.

**Comment by lassoan** [comment]
> Should sceneview changes be integrated in this PR ?

Related scene view changes were necessary (scene view restore did not work or crash).

**Comment by lassoan** [comment]
Thanks a lot for working on these @jcfr!

**Comment by jcfr** [comment]
> > Should sceneview changes be integrated in this PR ?

> Related scene view changes were necessary (scene view restore did not work or crash).

Or to say differently, while the scene view changes are important. Are they relevant to allow recording and playing of sequences ? 

If not, I suggest we integrate them separately.

**Comment by jcfr** [comment]
Nevermind, the sceneview tests will be fixed at later time ... as they do not work, fail or crash with the current set of changes.

**Comment by lassoan** [comment]
There were special hacks in node copy method just for preventing scene views from crashing/corrupting data. Those workarounds were not compatible with the new API (or it would have been messy and complicated to preserve). If we integrate all these commits at the same time then it does not matter much if they are in the exact same commit or in two separate commits.

As far as I rememver, I've fixed all sceneviews changes, but if not, it should be no problem, as I'll rework scene views completely anyway (to use sequences as storage backend).

**Comment by jcfr** [comment]
I will do a local build and then integrate. Commit message has been updated to include more details.

**Comment by jcfr** [comment]
Identified warnings - working on fixing them:

```
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation.cxx: In member function ‘bool vtkSlicerMarkupsWidgetRepresentation::IsDisplayable()’:
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation.cxx:760:23: warning: variable ‘effectiveDisplayNode’ set but not used [-Wunused-but-set-variable]
   vtkMRMLDisplayNode* effectiveDisplayNode = this->MarkupsDisplayNode;
                       ^
```

```
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx: In member function ‘virtual void vtkSlicerMarkupsWidgetRepresentation2D::UpdateFromMRML(vtkMRMLNode*, long unsigned int, void*)’:
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx:423:8: warning: variable ‘hierarchyVisibility’ set but not used [-Wunused-but-set-variable]
   bool hierarchyVisibility = true;
        ^
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx: In member function ‘virtual void vtkSlicerMarkupsWidgetRepresentation2D::CanInteractWithHandles(vtkMRMLInteractionEventData*, int&, int&, double&)’:
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx:633:14: warning: unused variable ‘dist2’ [-Wunused-variable]
       double dist2 = vtkMath::Distance2BetweenPoints(handleDisplayPos, displayPosition3);
              ^
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx: In member function ‘virtual bool vtkSlicerMarkupsWidgetRepresentation2D::IsCenterDisplayableOnSlice(vtkMRMLMarkupsNode*)’:
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidgetRepresentation2D.cxx:1040:23: warning: unused variable ‘displayNode’ [-Wunused-variable]
   vtkMRMLDisplayNode *displayNode = markupsNode->GetDisplayNode();
                       ^
```

```
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidget.cxx:1237:61: warning: unused parameter ‘type’ [-Wunused-parameter]
 bool vtkSlicerMarkupsWidget::GetIntersectionOnAxisPlane(int type, int index, const double input_Display[2], double outputIntersection_World[3])
                                                             ^
/home/jcfr/Projects/Slicer/Modules/Loadable/Markups/VTKWidgets/vtkSlicerMarkupsWidget.cxx:1327:67: warning: unused parameter ‘type’ [-Wunused-parameter]
 bool vtkSlicerMarkupsWidget::GetClosestPointOnInteractionAxis(int type, int index, const double input_Display[2], double outputClosestPoint_World[3])
                                                                   ^
```

```
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/Logic/vtkSlicerSequencesLogic.cxx: In member function ‘vtkMRMLSequenceNode* vtkSlicerSequencesLogic::AddSequence(const char*)’:
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/Logic/vtkSlicerSequencesLogic.cxx:168:12: warning: converting ‘false’ to pointer type ‘vtkMRMLSequenceNode*’ [-Wconversion-null]
     return false;
            ^
```


```
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/qSlicerSequencesModuleWidget.cxx: In constructor ‘qSlicerSequencesModuleWidgetPrivate::qSlicerSequencesModuleWidgetPrivate(qSlicerSequencesModuleWidget&)’:
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/qSlicerSequencesModuleWidget.cxx:118:11: warning: ‘qSlicerSequencesModuleWidgetPrivate::DataNodeCandidatesClassName’ will be initialized after [-Wreorder]
   QString DataNodeCandidatesClassName; // data node class name that was used for populating the candidate node list
           ^
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/qSlicerSequencesModuleWidget.cxx:103:8: warning:   ‘bool qSlicerSequencesModuleWidgetPrivate::ModuleWindowInitialized’ [-Wreorder]
   bool ModuleWindowInitialized;
        ^
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/qSlicerSequencesModuleWidget.cxx:130:1: warning:   when initialized here [-Wreorder]
 qSlicerSequencesModuleWidgetPrivate::qSlicerSequencesModuleWidgetPrivate( qSlicerSequencesModuleWidget& object )
 ^
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/qSlicerSequencesModuleWidget.cxx: At global scope:
/home/jcfr/Projects/Slicer/Modules/Loadable/Sequences/qSlicerSequencesModuleWidget.cxx:1570:108: warning: unused parameter ‘context’ [-Wunused-parameter]
 bool qSlicerSequencesModuleWidget::setEditedNode(vtkMRMLNode* node, QString role /* = QString()*/, QString context /* = QString() */)
                                                                                                            ^

```

**Comment by jcfr** [comment]
Got this error also:

```
/usr/bin/ld: cannot open output file /bin/designer/libqSlicerSequencesModuleWidgetsPlugins.so: No such file or directory
collect2: error: ld returned 1 exit status
Modules/Loadable/Sequences/Widgets/DesignerPlugins/CMakeFiles/qSlicerSequencesModuleWidgetsPlugins.dir/build.make:495: recipe for target '/bin/designer/libqSlicerSequencesModuleWidgetsPlugins.so' failed
make[5]: *** [/bin/designer/libqSlicerSequencesModuleWidgetsPlugins.so] Error 1

```

**Comment by lassoan** [comment]
For the link error: qSlicerSequenceBrowserModuleWidgets was renamed to qSlicerSequencesModuleWidgets, but maybe I missed some renames.

**Comment by jcfr** [comment]
>For the link error: qSlicerSequenceBrowserModuleWidgets was renamed to qSlicerSequencesModuleWidgets, but maybe I missed some renames.

In fact, this is related to https://github.com/Slicer/Slicer/pull/4810/files#diff-36d1fa83387ebcabaa3fcd7dd86eca1eR33-R35

Fix being tested

**Comment by jcfr** [comment]
Success: 

image

**Comment by jcfr** [comment]
Following issues:
* #4841: Improve SceneView to use sequences as storage backend and re-enable tests
* #4840: Add test for vtkMRMLNode::HasCopyContent

**Pull request merged**